<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Simple Mendelian randomisation examples â€¢ simulateGP</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Simple Mendelian randomisation examples">
<meta property="og:description" content="simulateGP">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">simulateGP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/index.html">Guide</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/explodecomputer/simulateGP">Source</a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="simplemr_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Simple Mendelian randomisation examples</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/explodecomputer/simulateGP/blob/master/vignettes/simplemr.Rmd"><code>vignettes/simplemr.Rmd</code></a></small>
      <div class="hidden name"><code>simplemr.Rmd</code></div>

    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit">
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://explodecomputer.github.io/simulateGP/">simulateGP</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="http://www.systemfit.org">systemfit</a></span><span class="op">)</span>
<span class="co">#&gt; Loading required package: Matrix</span>
<span class="co">#&gt; Loading required package: car</span>
<span class="co">#&gt; Loading required package: carData</span>
<span class="co">#&gt; Loading required package: lmtest</span>
<span class="co">#&gt; Loading required package: zoo</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'zoo'</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     as.Date, as.Date.numeric</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Please cite the 'systemfit' package as:</span>
<span class="co">#&gt; Arne Henningsen and Jeff D. Hamann (2007). systemfit: A Package for Estimating Systems of Simultaneous Equations in R. Journal of Statistical Software 23(4), 1-40. http://www.jstatsoft.org/v23/i04/.</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; If you have questions, suggestions, or comments regarding the 'systemfit' package, please use a forum or 'tracker' at systemfit's R-Forge site:</span>
<span class="co">#&gt; https://r-forge.r-project.org/projects/systemfit/</span></pre></div>
<p>The Mendelian randomisation statistical method aims to estimate the causal effect of some exposure <span class="math inline">\(x\)</span> on some outcome <span class="math inline">\(y\)</span> using a genetic instrumental variable for the exposure, <span class="math inline">\(g\)</span>. The assumptions of the model are that</p>
<ol style="list-style-type: decimal">
<li>
<span class="math inline">\(g\)</span> associates with <span class="math inline">\(x\)</span>
</li>
<li>
<span class="math inline">\(g\)</span> is independent of any confounders of <span class="math inline">\(x\)</span> and <span class="math inline">\(y\)</span>
</li>
<li>
<span class="math inline">\(g\)</span> only associates with <span class="math inline">\(y\)</span> via <span class="math inline">\(x\)</span>
</li>
</ol>
<p>A DAG representing the assumptions is below:</p>
<div class="figure">
<img src="mrdag.png" alt=""><p class="caption">mrdag</p>
</div>
<p>We can simulate individual level data according to this DAG</p>
<ol style="list-style-type: decimal">
<li>Simulate some genetic or confounding variables</li>
<li>Simulate exposures that are influenced by (1)</li>
<li>Simulate the outcomes that are influenced by (1) and (2)</li>
<li>Obtain MR estimate using two-stage least squares</li>
</ol>
<p>Here is how to do 1-3:</p>
<div class="sourceCode" id="cb2"><pre class="downlit">
<span class="co"># Set causal effect of x on y</span>
<span class="va">beta_xy</span> <span class="op">&lt;-</span> <span class="op">-</span><span class="fl">0.3</span>

<span class="co"># Set number of instruments for x</span>
<span class="va">nsnp</span> <span class="op">&lt;-</span> <span class="fl">3</span>

<span class="co"># Set number of individuals to simulate</span>
<span class="va">nid</span> <span class="op">&lt;-</span> <span class="fl">10000</span>

<span class="co"># Set variance explained in x by the instruments</span>
<span class="va">rsq_gx</span> <span class="op">&lt;-</span> <span class="fl">0.05</span>

<span class="co"># Generate a confounder</span>
<span class="va">u</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span><span class="op">(</span><span class="va">nid</span><span class="op">)</span>

<span class="co"># Generate genotypes with allele frequencies of 0.5</span>
<span class="va">g</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_geno.html">make_geno</a></span><span class="op">(</span>nid<span class="op">=</span><span class="va">nid</span>, nsnp<span class="op">=</span><span class="va">nsnp</span>, af<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span>

<span class="co"># These SNPs instrument some exposure, and together explain 5% of the variance</span>
<span class="va">effs</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/choose_effects.html">choose_effects</a></span><span class="op">(</span>nsnp<span class="op">=</span><span class="va">nsnp</span>, totvar<span class="op">=</span><span class="va">rsq_gx</span><span class="op">)</span>

<span class="co"># Create X - influenced by snps and the confounder</span>
<span class="va">x</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span>effs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">effs</span>, <span class="fl">0.3</span><span class="op">)</span>, indep<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">g</span>, <span class="va">u</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>Check that the SNPs explain 5% of the variance in x</p>
<div class="sourceCode" id="cb3"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/ieugwasr/man/cor.html">cor</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">g</span><span class="op">)</span><span class="op">^</span><span class="fl">2</span><span class="op">)</span>
<span class="co">#&gt; [1] 0.0511963</span></pre></div>
<p>Create Y, which is negatively influenced by x and positively influenced by the confounder</p>
<div class="sourceCode" id="cb4"><pre class="downlit">
<span class="va">y</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/make_phen.html">make_phen</a></span><span class="op">(</span>effs<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">beta_xy</span>, <span class="fl">0.3</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/cbind.html">cbind</a></span><span class="op">(</span><span class="va">x</span>, <span class="va">u</span><span class="op">)</span><span class="op">)</span></pre></div>
<p>We now have an X and Y, and the genotypes. To perform 2-stage least squares MR on this we can use the <code>systemfit</code> package.</p>
<div class="sourceCode" id="cb5"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu">systemfit</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/systemfit/man/systemfit.html">systemfit</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span>, method<span class="op">=</span><span class="st">"2SLS"</span>, inst <span class="op">=</span> <span class="op">~</span> <span class="va">g</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; systemfit results </span>
<span class="co">#&gt; method: 2SLS </span>
<span class="co">#&gt; </span>
<span class="co">#&gt;            N   DF     SSR  detRCov   OLS-R2 McElroy-R2</span>
<span class="co">#&gt; system 10000 9998 9528.58 0.953049 0.047047   0.047047</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;         N   DF     SSR      MSE     RMSE       R2   Adj R2</span>
<span class="co">#&gt; eq1 10000 9998 9528.58 0.953049 0.976242 0.047047 0.046951</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; The covariance matrix of the residuals</span>
<span class="co">#&gt;          eq1</span>
<span class="co">#&gt; eq1 0.953049</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; The correlations of the residuals</span>
<span class="co">#&gt;     eq1</span>
<span class="co">#&gt; eq1   1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; 2SLS estimates for 'eq1' (equation 1)</span>
<span class="co">#&gt; Model Formula: y ~ x</span>
<span class="co">#&gt; Instruments: ~g</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;                 Estimate   Std. Error  t value   Pr(&gt;|t|)    </span>
<span class="co">#&gt; (Intercept) -6.72797e-18  9.76242e-03  0.00000          1    </span>
<span class="co">#&gt; x           -2.27575e-01  4.33722e-02 -5.24703 1.5774e-07 ***</span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Residual standard error: 0.976242 on 9998 degrees of freedom</span>
<span class="co">#&gt; Number of observations: 10000 Degrees of Freedom: 9998 </span>
<span class="co">#&gt; SSR: 9528.57923 MSE: 0.953049 Root MSE: 0.976242 </span>
<span class="co">#&gt; Multiple R-Squared: 0.047047 Adjusted R-Squared: 0.046951</span></pre></div>
<p>Compare against confounded observational estimate</p>
<div class="sourceCode" id="cb6"><pre class="downlit">
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/lm.html">lm</a></span><span class="op">(</span><span class="va">y</span> <span class="op">~</span> <span class="va">x</span><span class="op">)</span><span class="op">)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Call:</span>
<span class="co">#&gt; lm(formula = y ~ x)</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Residuals:</span>
<span class="co">#&gt;     Min      1Q  Median      3Q     Max </span>
<span class="co">#&gt; -3.9817 -0.6614 -0.0137  0.6659  3.9991 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Coefficients:</span>
<span class="co">#&gt;               Estimate Std. Error t value Pr(&gt;|t|)    </span>
<span class="co">#&gt; (Intercept) -7.671e-17  9.762e-03    0.00        1    </span>
<span class="co">#&gt; x           -2.172e-01  9.762e-03  -22.24   &lt;2e-16 ***</span>
<span class="co">#&gt; ---</span>
<span class="co">#&gt; Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Residual standard error: 0.9762 on 9998 degrees of freedom</span>
<span class="co">#&gt; Multiple R-squared:  0.04716,    Adjusted R-squared:  0.04706 </span>
<span class="co">#&gt; F-statistic: 494.8 on 1 and 9998 DF,  p-value: &lt; 2.2e-16</span></pre></div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by Gibran Hemani, John Ferguson.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
