<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Function reference â€¢ simulateGP</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/united/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Function reference"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-index">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">simulateGP</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../articles/index.html">Guide</a>
</li>
<li>
  <a href="../reference/index.html">Functions</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/explodecomputer/simulateGP" class="external-link">Source</a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
    <div class="page-header">
      <h1>Reference</h1>
    </div>

    <table class="ref-index"><colgroup><col class="alias"><col class="title"></colgroup><tbody><tr><th colspan="2">
          <h2 id="all-functions">All functions <a href="#all-functions" class="anchor" aria-hidden="true"></a></h2>
          <p class="section-desc"></p>
        </th>
      </tr></tbody><tbody><tr><td>
          <p><code><a href="add_u.html">add_u()</a></code> </p>
        </td>
        <td><p>Add confounder variables and their instruments</p></td>
      </tr><tr><td>
          <p><code><a href="allele_frequency.html">allele_frequency()</a></code> </p>
        </td>
        <td><p>Estimate allele frequency from SNP</p></td>
      </tr><tr><td>
          <p><code><a href="arbitrary_map.html">arbitrary_map()</a></code> </p>
        </td>
        <td><p>Create an arbitrary map</p></td>
      </tr><tr><td>
          <p><code><a href="ascertain_samples.html">ascertain_samples()</a></code> </p>
        </td>
        <td><p>Ascertain some proportion of cases and controls from binary phenotype</p></td>
      </tr><tr><td>
          <p><code><a href="choose_effects.html">choose_effects()</a></code> </p>
        </td>
        <td><p>Get vector of effects that explain some amount of variance</p></td>
      </tr><tr><td>
          <p><code><a href="contingency.html">contingency()</a></code> </p>
        </td>
        <td><p>Obtain 2x2 contingency table from marginal parameters and odds ratio</p></td>
      </tr><tr><td>
          <p><code><a href="correlated_binomial.html">correlated_binomial()</a></code> </p>
        </td>
        <td><p>Simulate two correlated binomial variables</p></td>
      </tr><tr><td>
          <p><code><a href="create_system.html">create_system()</a></code> </p>
        </td>
        <td><p>Wrapper for simulation pipeline</p></td>
      </tr><tr><td>
          <p><code><a href="estimate_system_effects.html">estimate_system_effects()</a></code> </p>
        </td>
        <td><p>Estimate the effects of all SNPs on all phenotypes</p></td>
      </tr><tr><td>
          <p><code><a href="expected_mse.html">expected_mse()</a></code> </p>
        </td>
        <td><p>Calculate expected MSE</p></td>
      </tr><tr><td>
          <p><code><a href="expected_se.html">expected_se()</a></code> </p>
        </td>
        <td><p>Expected se given beta, af, n and vy</p></td>
      </tr><tr><td>
          <p><code><a href="expected_ssx.html">expected_ssx()</a></code> </p>
        </td>
        <td><p>Calculate expected SSX</p></td>
      </tr><tr><td>
          <p><code><a href="fast_assoc.html">fast_assoc()</a></code> </p>
        </td>
        <td><p>Get summary statistics in simple linear regression</p></td>
      </tr><tr><td>
          <p><code><a href="generate_gwas_params.html">generate_gwas_params()</a></code> </p>
        </td>
        <td><p>Generate SNP effects given MAF, h2 and selection</p></td>
      </tr><tr><td>
          <p><code><a href="generate_gwas_ss.html">generate_gwas_ss()</a></code> </p>
        </td>
        <td><p>Modify SNP effects to account for LD</p></td>
      </tr><tr><td>
          <p><code><a href="generate_gwas_ss_1.html">generate_gwas_ss_1()</a></code> </p>
        </td>
        <td><p>Create a GWAS summary dataset</p></td>
      </tr><tr><td>
          <p><code><a href="generate_ldobj.html">generate_ldobj()</a></code> </p>
        </td>
        <td><p>Generate LD matrix objects from reference panel</p></td>
      </tr><tr><td>
          <p><code><a href="get_effs.html">get_effs()</a></code> </p>
        </td>
        <td><p>Get effs for two traits and make dat format</p></td>
      </tr><tr><td>
          <p><code><a href="get_ld.html">get_ld()</a></code> </p>
        </td>
        <td><p>Get LD matrix for a specified region from bfile reference panel</p></td>
      </tr><tr><td>
          <p><code><a href="get_population_allele_frequency.html">get_population_allele_frequency()</a></code> </p>
        </td>
        <td><p>Estimate the allele frequency in population from case/control summary data</p></td>
      </tr><tr><td>
          <p><code><a href="get_regions_from_ldobjdir.html">get_regions_from_ldobjdir()</a></code> </p>
        </td>
        <td><p>Determine regions from LD file</p></td>
      </tr><tr><td>
          <p><code><a href="gwas.html">gwas()</a></code> </p>
        </td>
        <td><p>Perform association of many SNPs against phenotype</p></td>
      </tr><tr><td>
          <p><code><a href="gx_to_gp.html">gx_to_gp()</a></code> </p>
        </td>
        <td><p>Translate risk from liability to probability scale</p></td>
      </tr><tr><td>
          <p><code><a href="hap_freqs.html">hap_freqs()</a></code> </p>
        </td>
        <td><p>Estimate haplotype frequencies for two loci</p></td>
      </tr><tr><td>
          <p><code><a href="init_parameters.html">init_parameters()</a></code> </p>
        </td>
        <td><p>Choose initial parameters for direct effects on X and Y</p></td>
      </tr><tr><td>
          <p><code><a href="ldetect.html">ldetect</a></code> </p>
        </td>
        <td><p>Data frame of independent LD regions</p></td>
      </tr><tr><td>
          <p><code><a href="lor_to_rsq.html">lor_to_rsq()</a></code> </p>
        </td>
        <td><p>Estimate proportion of variance of liability explained by SNP in general population</p></td>
      </tr><tr><td>
          <p><code><a href="make_geno.html">make_geno()</a></code> </p>
        </td>
        <td><p>Create genotype matrix</p></td>
      </tr><tr><td>
          <p><code><a href="make_mvdat.html">make_mvdat()</a></code> </p>
        </td>
        <td><p>Take several exposures and one outcome and make the data required for multivariable MR</p></td>
      </tr><tr><td>
          <p><code><a href="make_phen.html">make_phen()</a></code> </p>
        </td>
        <td><p>Simulate variable based on the influences of other variables</p></td>
      </tr><tr><td>
          <p><code><a href="merge_exp_out.html">merge_exp_out()</a></code> </p>
        </td>
        <td><p>Organise outputs from <code>gwas</code> into harmonised dat format</p></td>
      </tr><tr><td>
          <p><code><a href="range01.html">range01()</a></code> </p>
        </td>
        <td><p>Scale variable to have range between 0 and 1</p></td>
      </tr><tr><td>
          <p><code><a href="read_ldobjdir.html">read_ldobjdir()</a></code> </p>
        </td>
        <td><p>Read in LD objects into list</p></td>
      </tr><tr><td>
          <p><code><a href="recode_dat.html">recode_dat()</a></code> </p>
        </td>
        <td><p>Recode data to make every effect on x positive</p></td>
      </tr><tr><td>
          <p><code><a href="recode_dat_intercept.html">recode_dat_intercept()</a></code> </p>
        </td>
        <td><p>Intercept recoding to have every effect on x positive</p></td>
      </tr><tr><td>
          <p><code><a href="recode_dat_simple.html">recode_dat_simple()</a></code> </p>
        </td>
        <td><p>Simple recoding to have every effect on x positive</p></td>
      </tr><tr><td>
          <p><code><a href="risk_cross_plot.html">risk_cross_plot()</a></code> </p>
        </td>
        <td><p>Plot liability vs probability disease risk</p></td>
      </tr><tr><td>
          <p><code><a href="risk_simulation.html">risk_simulation()</a></code> </p>
        </td>
        <td><p>Make simulation to compare disease and liability scales</p></td>
      </tr><tr><td>
          <p><code><a href="sample_beta.html">sample_beta()</a></code> </p>
        </td>
        <td><p>Sample beta values given standard errors</p></td>
      </tr><tr><td>
          <p><code><a href="sample_system_effects.html">sample_system_effects()</a></code> </p>
        </td>
        <td><p>Sample the actual effects based on initial parameters</p></td>
      </tr><tr><td>
          <p><code><a href="simulateGP-package.html">simulateGP-package</a></code> <code><a href="simulateGP-package.html">simulateGP</a></code> <code><a href="simulateGP-package.html">simulategp</a></code> </p>
        </td>
        <td><p>simulateGP: Functions for Simulating Genotype-Phenotype Relationships</p></td>
      </tr><tr><td>
          <p><code><a href="simulate_geno.html">simulate_geno()</a></code> </p>
        </td>
        <td><p>Simulate genotypes from haplotypes</p></td>
      </tr><tr><td>
          <p><code><a href="simulate_haplotypes.html">simulate_haplotypes()</a></code> </p>
        </td>
        <td><p>Simulate haplotypes of two loci</p></td>
      </tr><tr><td>
          <p><code><a href="simulate_population.html">simulate_population()</a></code> </p>
        </td>
        <td><p>Simulate individual level data from initial parameters</p></td>
      </tr><tr><td>
          <p><code><a href="stuff.html">stuff</a></code> </p>
        </td>
        <td><p>General funcs</p></td>
      </tr><tr><td>
          <p><code><a href="summary_set.html">summary_set()</a></code> </p>
        </td>
        <td><p>Wrapper for generating a summary set</p></td>
      </tr><tr><td>
          <p><code><a href="test_ldobj.html">test_ldobj()</a></code> </p>
        </td>
        <td><p>Create test LD object</p></td>
      </tr><tr><td>
          <p><code><a href="test_system.html">test_system()</a></code> </p>
        </td>
        <td><p>Apply MR tests to system</p></td>
      </tr><tr><td>
          <p><code><a href="y_to_binary.html">y_to_binary()</a></code> </p>
        </td>
        <td><p>Convert continuous trait to binary</p></td>
      </tr></tbody></table></div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Gibran Hemani, John Ferguson, Rita Rasteiro.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

